### DISCUSSION TITLE
Header configuration for all pages but not APIs

### DISCUSSION THREAD

**Discussion by paambaati** [vercel/next.js#16768] [Help] [answered]
I'm trying to [add a few headers to my Next.js config](https://nextjs.org/docs/api-reference/next.config.js/headers) but I'm having trouble making them work only for all pages but not API routes.

Here's what I've tried so far —

1. This works for __all__ routes, include APIs.

```javascript
async headers() {
    return [{
        source: '/(.*)',
        headers: [{
            key: 'X-Custom-Page-Only-Header',
            value: 'ok',
        }]
    }]
}
```

2. Negative look-aheads don't seem to work and don't throw any errors.

```javascript
source: '/(?!^\/api\/$)'
```

**Comment by ijjk** [comment, marked as answer]
Hi, you should be seeing an error with the above `source` (`'/(?!^\/api\/$)'`) what version of Next.js are you on? Using the above source the following error is shown on the latest version of Next.js v9.5.3

```sh
Error parsing `/(?!^/api/$)` https://err.sh/vercel/next.js/invalid-route-source
Reason: Pattern cannot start with "?" at 2

  /(?!^/api/$)
    ^

`source` parse failed for route {"source":"/(?!^/api/$)","headers":[{"key":"x-hello","value":"world"}]}

Error: Invalid header found
```

To match against all routes except `/api/*` you would want the something like the below route.

```js
module.exports = {
  headers() {
    return [
      {
        source: '/((?!api$|api/).*)',
        headers: [
          {
            key: 'X-Custom-Page-Only-Header',
            value: 'ok'
          }
        ]
      }
    ]
  }
}
```

**Reply by paambaati** [reply]
Thanks, this works @ijjk! And you're right, I did see that error; I was trying out a bunch of regexes and must've not noticed which one threw this error.

**Reply by StephaneBischoff** [reply]
For me, this worked better. My root url was not being catched.
`'/:path((?!api).*)*',`
